diff --git a/_everything.scss b/_everything.scss
index 6069365ca..56fa50cf7 100644
--- a/_everything.scss
+++ b/_everything.scss
@@ -15,5 +15,9 @@
       padding: 2rem,
       gap: 0.5rem,
     ),
-  )
+  ),
+  // $form-disable-textarea-resizing-rows: true,
+  // $form-disable-text-field-filled-theme: true,
+  // $form-disable-text-field-underlined-theme: true,
+  // $form-disable-text-field-outlined-theme: true
 );
diff --git a/packages/form/src/index.ts b/packages/form/src/index.ts
index ea6e67ac7..f81ae8dbd 100644
--- a/packages/form/src/index.ts
+++ b/packages/form/src/index.ts
@@ -15,4 +15,7 @@ export * from "./TextArea";
 export * from "./TextField";
 export * from "./TextFieldContainer";
 export * from "./types";
+export * from "./useConfirmedPassword";
 export * from "./useFileUpload";
+export * from "./useTextField";
+export * from "./validation";
diff --git a/packages/form/src/useConfirmedPassword.ts b/packages/form/src/useConfirmedPassword.ts
new file mode 100644
index 000000000..92d0165d8
--- /dev/null
+++ b/packages/form/src/useConfirmedPassword.ts
@@ -0,0 +1,221 @@
+import { ProvidedTextFieldMessageProps, useTextField } from "./useTextField";
+import {
+  defaultGetErrorMessage,
+  ErrorMessageOptions,
+  GetErrorMessage,
+} from "./validation";
+
+export interface PasswordErrorMessageOptions
+  extends Omit<ErrorMessageOptions, "value"> {
+  password: string;
+  confirmed: string;
+  valueMissing: string;
+  valueMismatch: string;
+  patternMismatch: string;
+}
+
+export type GetPasswordErrorMessage = (
+  options: PasswordErrorMessageOptions
+) => string;
+
+// const defaultGetPasswordErrorMessage: GetPasswordErrorMessage = (options) => {
+//   const {
+//     password,
+//     confirmed,
+//     validity,
+//     valueMissing,
+//     valueMismatch,
+//     patternMismatch,
+//   } = options;
+//   if (validity.valueMissing && valueMissing) {
+//     return valueMissing;
+//   }
+
+//   if (validity.patternMismatch && patternMismatch) {
+//     return patternMismatch;
+//   }
+
+//   if (confirmed && password !== confirmed) {
+//     return valueMismatch;
+//   }
+
+//   return "";
+// };
+
+export interface ConfirmedPasswordOptions {
+  pattern?: string;
+  required?: boolean;
+  errorIcon?: boolean;
+  valueMissing?: string;
+  valueMismatch?: string;
+  patternMismatch?: string;
+  getErrorMessage?: GetPasswordErrorMessage;
+}
+
+export interface ConfirmedPasswordHookReturnValue {
+  error: boolean;
+  passwordProps: ProvidedTextFieldMessageProps & { name: "password" };
+  confirmPasswordProps: ProvidedTextFieldMessageProps & {
+    name: "confirmPassword";
+  };
+}
+
+export function useConfirmedPassword(
+  options: ConfirmedPasswordOptions = {}
+): ConfirmedPasswordHookReturnValue {
+  const {
+    pattern,
+    required = true,
+    errorIcon = false,
+    valueMissing = "This value is required.",
+    valueMismatch = "Passwords do not match.",
+    patternMismatch = "",
+    // getErrorMessage = defaultGetPasswordErrorMessage,
+  } = options;
+  const createGetErrorMessage =
+    (isConfirm: boolean): GetErrorMessage =>
+    (options) => {
+      const { validity } = options;
+      if (validity.valueMissing) {
+        return valueMissing;
+      }
+
+      if (validity.patternMismatch && patternMismatch) {
+        return patternMismatch;
+      }
+
+      const currentPassword = isConfirm ? password : options.value;
+      const currentConfirm = isConfirm ? options.value : confirmPassword;
+      if (currentConfirm && currentConfirm !== currentPassword) {
+        return valueMismatch;
+      }
+
+      return defaultGetErrorMessage(options);
+    };
+
+  const getPasswordError = createGetErrorMessage(false);
+  const getConfirmError = createGetErrorMessage(true);
+  // const getPasswordError: GetErrorMessage = (options) =>
+  //   getErrorMessage({
+  //     ...options,
+  //     password,
+  //     confirmed,
+  //     valueMismatch,
+  //     valueMissing,
+  //     patternMismatch,
+  //   });
+  // const getConfirmError: GetErrorMessage = (options) =>
+  //   getErrorMessage({
+  //     ...options,
+  //     password,
+  //     confirmed,
+  //     valueMismatch,
+  //     valueMissing,
+  //     patternMismatch,
+  //   });
+
+  const {
+    value: password,
+    error: passwordError,
+    setState: setPasswordState,
+    fieldRef: passwordRef,
+    fieldProps: passwordProps,
+  } = useTextField({
+    pattern,
+    required,
+    errorIcon,
+    getErrorMessage: getPasswordError,
+    onChange(event) {
+      const { value, validity, validationMessage } = event.currentTarget;
+      const errorMessage = getPasswordError({
+        value,
+        pattern,
+        validity,
+        validationMessage:
+          validationMessage === valueMismatch ? "" : validationMessage,
+        validateOnChange: true,
+        isBlurEvent: false,
+      });
+
+      setConfirmPasswordState((prevState) => {
+        if (errorMessage === valueMismatch) {
+          return {
+            ...prevState,
+            error: true,
+            errorMessage,
+          };
+        }
+
+        if (prevState.errorMessage === valueMismatch && !errorMessage) {
+          confirmPasswordRef.current?.setCustomValidity("");
+          return {
+            ...prevState,
+            error: false,
+            errorMessage: "",
+          };
+        }
+
+        return prevState;
+      });
+    },
+  });
+
+  const {
+    value: confirmPassword,
+    error: confirmPasswordError,
+    setState: setConfirmPasswordState,
+    fieldRef: confirmPasswordRef,
+    fieldProps: confirmPasswordProps,
+  } = useTextField({
+    pattern,
+    required,
+    errorIcon,
+    getErrorMessage: getConfirmError,
+    onChange(event) {
+      const { value, validity, validationMessage } = event.currentTarget;
+      const errorMessage = getConfirmError({
+        value,
+        pattern,
+        validity,
+        validationMessage:
+          validationMessage === valueMismatch ? "" : validationMessage,
+        validateOnChange: true,
+        isBlurEvent: false,
+      });
+
+      setPasswordState((prevState) => {
+        if (errorMessage === valueMismatch) {
+          return {
+            ...prevState,
+            error: true,
+            errorMessage,
+          };
+        }
+
+        if (prevState.errorMessage === valueMismatch && !errorMessage) {
+          passwordRef.current?.setCustomValidity("");
+          return {
+            ...prevState,
+            error: false,
+            errorMessage: "",
+          };
+        }
+
+        return prevState;
+      });
+    },
+  });
+
+  return {
+    error:
+      !password || !confirmPassword || !passwordError || !confirmPasswordError,
+    passwordProps: {
+      ...passwordProps,
+      name: "password",
+    },
+    confirmPasswordProps: {
+      ...confirmPasswordProps,
+      name: "confirmPassword",
+    },
+  };
+}
diff --git a/packages/form/src/useTextField.ts b/packages/form/src/useTextField.ts
new file mode 100644
index 000000000..3fe9f7659
--- /dev/null
+++ b/packages/form/src/useTextField.ts
@@ -0,0 +1,455 @@
+import {
+  useEnsuredId,
+  useEnsuredRef,
+  UseStateInitializer,
+  UseStateSetter,
+} from "@react-md/core";
+import { useIcon } from "@react-md/icon";
+import {
+  FocusEvent,
+  ChangeEvent,
+  HTMLAttributes,
+  ReactNode,
+  useState,
+  useCallback,
+  useRef,
+  RefObject,
+  Ref,
+  RefCallback,
+} from "react";
+import {
+  FormMessageInputLengthCounterProps,
+  FormMessageProps,
+} from "./FormMessage";
+import { TextFieldProps } from "./TextField";
+import {
+  ChangeValidationBehavior,
+  ErrorMessageOptions,
+  GetErrorIcon,
+  GetErrorMessage,
+  IsErrored,
+  TextConstraints,
+  defaultIsErrored,
+  defaultGetErrorIcon,
+  defaultGetErrorMessage,
+} from "./validation";
+
+const noop = (): void => {
+  // do nothing
+};
+
+/**
+ * @remarks \@since 2.5.0
+ */
+export type TextFieldChangeHandlers = Pick<
+  HTMLAttributes<HTMLInputElement | HTMLTextAreaElement>,
+  "onBlur" | "onChange"
+>;
+
+/**
+ * A function that reports the error state changing. A good use-case for this is
+ * to keep track of all the errors within your form and keep a submit button
+ * disabled until they have been resolved.
+ *
+ * Example:
+ *
+ * ```ts
+ * const [errors, setErrors] = useState<Record<string, boolean | undefined>>({});
+ * const onErrorChange: ErrorChangeHandler = (id, error) =>
+ *   setErrors((prevErrors) => ({ ...prevErrors, [id]: error }));
+ *
+ * const invalid = Object.values(errors).some(Boolean);
+ *
+ * // form implementation is left as an exercise for the reader
+ * <Button type="submit" disabled={invalid} onClick={submitForm}>Submit</Button>
+ * ```
+ *
+ * @remarks \@since 2.5.0
+ */
+export type ErrorChangeHandler = (id: string, error: boolean) => void;
+
+/**
+ * @remarks \@since 2.5.0
+ */
+export interface TextFieldHookOptions
+  extends TextConstraints,
+    TextFieldChangeHandlers,
+    Pick<TextFieldProps, "id" | "theme"> {
+  ref?: Ref<HTMLInputElement | HTMLTextAreaElement>;
+
+  /**
+   * Boolean if the `FormMessage` should also display a counter for the
+   * remaining letters allowed based on the `maxLength`.
+   *
+   * This will still be considered false if the `maxLength` value is not
+   * provided.
+   *
+   * @defaultValue `false`
+   */
+  counter?: boolean;
+
+  /**
+   * The default value to use for the `TextField` or `TextArea` one initial
+   * render. If you want to manually change the value to something else after
+   * the initial render, either change the `key` for the component containing
+   * this hook, or use the `setState` function returned from this hook.
+   *
+   * @defaultValue `""`
+   */
+  defaultValue?: UseStateInitializer<string>;
+
+  /**
+   * An optional help text to display in the `FormMessage` component when there
+   * is not an error.
+   */
+  helpText?: ReactNode;
+
+  /**
+   * A function used to determine if the `TextField` or `TextArea` is an in
+   * errored state. See {@link defaultIsErrored} for the default implementation
+   * details.
+   */
+  isErrored?: IsErrored;
+
+  /**
+   * An optional error icon used in the {@link getErrorIcon} option.
+   */
+  errorIcon?: ReactNode;
+
+  /**
+   * A function used to get the error icon to display at the right of the
+   * `TextField` or `TextArea`. The default behavior will only show an icon when
+   * the `error` state is `true` and an `errorIcon` option has been provided.
+   */
+  getErrorIcon?: GetErrorIcon;
+
+  /**
+   * A function to get and display an error message based on the `TextField` or
+   * `TextArea` validity. See {@link defaultGetErrorMessage} for the default
+   * implementation details.
+   */
+  getErrorMessage?: GetErrorMessage;
+
+  /**
+   * An optional function that will be called whenever the `error` state is
+   * changed. This can be used for more complex forms to `disable` the Submit
+   * button or anything else if any field has an error.
+   */
+  onErrorChange?: ErrorChangeHandler;
+
+  /**
+   * Describes the validation behavior that should be done when the value within
+   * the `TextField` changes. This can either be:
+   *
+   * - a boolean
+   * - the string `"recommended"`
+   * - a single key of the ValidityState that should trigger the validation
+   * - a list of keys of the ValidityState that should trigger the validation
+   */
+  validateOnChange?: ChangeValidationBehavior;
+
+  /**
+   * Boolean if the `TextField` or `TextArea` will **not** be rendered along
+   * with a `FormMessage` component. This will prevent the `aria-describedby`
+   * prop from being returned when set to `true`.
+   *
+   * @defaultValue `false`
+   */
+  disableMessage?: boolean;
+
+  /**
+   * Boolean if the `maxLength` prop should not be passed to the `TextField`
+   * component since it will prevent any additional characters from being
+   * entered in the text field which might feel like weird behavior to some
+   * users. This should really only be used when the `counter` option is also
+   * enabled and rendering along with a `FormMessage` component.
+   *
+   * @defaultValue `false`
+   */
+  disableMaxLength?: boolean;
+}
+
+/**
+ * All the props that will be generated and return from the `useTextField` hook
+ * that should be passed to a `FormMessage` component.
+ *
+ * @remarks \@since 2.5.0
+ */
+export interface ProvidedFormMessageProps
+  extends Pick<FormMessageProps, "id" | "theme" | "children">,
+    Required<Pick<TextFieldProps, "error">>,
+    Partial<Pick<FormMessageInputLengthCounterProps, "length" | "maxLength">> {}
+
+/**
+ * All the props that will be generated and returned by the `useTextField` hook
+ * that should be passed to a `TextField` component.
+ *
+ * @remarks \@since 2.5.0
+ */
+export interface ProvidedTextFieldProps
+  extends TextConstraints,
+    Required<TextFieldChangeHandlers>,
+    Required<Pick<TextFieldProps, "id" | "value" | "error">>,
+    Pick<TextFieldProps, "aria-describedby" | "theme" | "rightAddon"> {
+  ref: RefCallback<HTMLInputElement | HTMLTextAreaElement>;
+}
+
+/**
+ * @remarks \@since 2.5.0
+ */
+export interface ProvidedTextFieldMessageProps extends ProvidedTextFieldProps {
+  /**
+   * These props will be defined as long as the `disableMessage` prop is not
+   * `true` from the `useTextField` hook.
+   */
+  messageProps: ProvidedFormMessageProps;
+}
+
+/**
+ * @remarks \@since 2.5.0
+ */
+export interface TextFieldHookState {
+  /**
+   * The current value for the `TextField` or `TextArea`.
+   */
+  value: string;
+
+  /**
+   * Boolean if the `TextField` or `TextArea` are in an errored state.
+   *
+   * Note: This can be `true` while the `errorMessage` is an empty string since
+   * it's sometimes useful to show there's an error while the user is typing
+   * without reporting the full error message.
+   */
+  error: boolean;
+
+  /**
+   * The current error message that will be an empty string when there are no
+   * errors for the `TextField` or `TextArea` component or the message is hidden
+   * during change events.
+   */
+  errorMessage: string;
+}
+
+export interface TextFieldHookReturnValue {
+  error: boolean;
+  value: string;
+  reset(): void;
+  checkValidity(isBlurEvent?: boolean): void;
+  setState: UseStateSetter<TextFieldHookState>;
+  fieldRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
+  fieldProps: ProvidedTextFieldProps | ProvidedTextFieldMessageProps;
+}
+
+export function useTextField(
+  options?: TextFieldHookOptions
+): TextFieldHookReturnValue & { fieldProps: ProvidedTextFieldMessageProps };
+export function useTextField(
+  options: TextFieldHookOptions & { disableMessage: true }
+): TextFieldHookReturnValue & { fieldProps: ProvidedTextFieldProps };
+export function useTextField(
+  options: TextFieldHookOptions = {}
+): TextFieldHookReturnValue {
+  const {
+    id: propId,
+    ref,
+    defaultValue = "",
+    theme,
+    pattern,
+    required,
+    minLength,
+    maxLength,
+    disableMaxLength = false,
+    onBlur,
+    onChange,
+    helpText,
+    errorIcon: propErrorIcon,
+    counter = false,
+    disableMessage = false,
+    validateOnChange = "recommended",
+    isErrored = defaultIsErrored,
+    onErrorChange = noop,
+    getErrorIcon = defaultGetErrorIcon,
+    getErrorMessage = defaultGetErrorMessage,
+  } = options;
+  const id = useEnsuredId(propId, "text-field");
+
+  const [state, setState] = useState<TextFieldHookState>(() => {
+    const value =
+      typeof defaultValue === "function" ? defaultValue() : defaultValue;
+
+    return {
+      value,
+      error: false,
+      errorMessage: "",
+    };
+  });
+
+  const { value, error, errorMessage } = state;
+  const errored = useRef(false);
+  const [fieldRef, fieldRefCallback] = useEnsuredRef(ref);
+
+  const messageId = `${id}-message`;
+  const isCounting = counter && typeof maxLength === "number";
+  const reset = useCallback(() => {
+    /* istanbul ignore next */
+    if (fieldRef.current) {
+      fieldRef.current.setCustomValidity("");
+    }
+
+    setState({
+      value: "",
+      error: false,
+      errorMessage: "",
+    });
+  }, [fieldRef]);
+
+  const checkValidity = useCallback(
+    (isBlurEvent = true) => {
+      const field = fieldRef.current;
+      if (!field) {
+        throw new Error("Unable to check validity due to missing ref");
+      }
+
+      // need to temporarily set the `maxLength` back so it can be "verified"
+      // through the validity api
+      /* istanbul ignore next */
+      if (isBlurEvent && disableMaxLength && typeof maxLength === "number") {
+        field.maxLength = maxLength;
+      }
+
+      const { value } = field;
+      field.setCustomValidity("");
+      field.checkValidity();
+
+      // remove the temporarily set `maxLength` attribute after checking the
+      // validity
+      /* istanbul ignore next */
+      if (disableMaxLength && typeof maxLength === "number") {
+        field.removeAttribute("maxLength");
+      }
+
+      const options: ErrorMessageOptions = {
+        value,
+        pattern,
+        required,
+        minLength,
+        maxLength,
+        isBlurEvent,
+        validity: field.validity,
+        validationMessage: field.validationMessage,
+        validateOnChange,
+      };
+      const errorMessage = getErrorMessage(options);
+      const error = isErrored({ ...options, errorMessage });
+
+      if (errored.current !== error) {
+        errored.current = error;
+        onErrorChange(id, error);
+      }
+
+      /* istanbul ignore next */
+      if (errorMessage !== field.validationMessage) {
+        field.setCustomValidity(errorMessage);
+      }
+
+      setState({ value, error, errorMessage });
+    },
+    [
+      disableMaxLength,
+      fieldRef,
+      getErrorMessage,
+      id,
+      isErrored,
+      maxLength,
+      minLength,
+      onErrorChange,
+      pattern,
+      required,
+      validateOnChange,
+    ]
+  );
+
+  const handleBlur = useCallback(
+    (event: FocusEvent<HTMLInputElement>) => {
+      if (onBlur) {
+        onBlur(event);
+      }
+
+      if (event.isPropagationStopped()) {
+        return;
+      }
+
+      checkValidity(true);
+    },
+    [onBlur, checkValidity]
+  );
+  const handleChange = useCallback(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      if (onChange) {
+        onChange(event);
+      }
+
+      if (event.isPropagationStopped()) {
+        return;
+      }
+
+      if (
+        !validateOnChange ||
+        (Array.isArray(validateOnChange) && !validateOnChange.length)
+      ) {
+        setState((prevState) => ({
+          ...prevState,
+          value: event.currentTarget.value,
+        }));
+        return;
+      }
+
+      checkValidity(false);
+    },
+    [onChange, checkValidity, validateOnChange]
+  );
+
+  const errorIcon = useIcon("error", propErrorIcon);
+  const fieldProps: ProvidedTextFieldProps & {
+    messageProps?: ProvidedFormMessageProps;
+  } = {
+    id,
+    ref: fieldRefCallback,
+    value,
+    theme,
+    error,
+    required,
+    pattern,
+    minLength,
+    maxLength: disableMaxLength ? undefined : maxLength,
+    rightAddon: getErrorIcon({
+      error,
+      errorIcon,
+      errorMessage,
+    }),
+    onBlur: handleBlur,
+    onChange: handleChange,
+  };
+  if (!disableMessage) {
+    fieldProps["aria-describedby"] = messageId;
+    fieldProps.messageProps = {
+      id: messageId,
+      error,
+      theme,
+      length: counter ? value.length : undefined,
+      maxLength: isCounting ? maxLength : undefined,
+      children: errorMessage || helpText,
+    };
+  }
+
+  return {
+    value,
+    error,
+    reset,
+    setState,
+    fieldRef,
+    fieldProps,
+    checkValidity,
+  };
+}
diff --git a/packages/form/src/validation.ts b/packages/form/src/validation.ts
new file mode 100644
index 000000000..b5decad75
--- /dev/null
+++ b/packages/form/src/validation.ts
@@ -0,0 +1,259 @@
+import type { InputHTMLAttributes, ReactNode } from "react";
+
+/**
+ * @remarks \@since 2.5.0
+ */
+export type TextConstraints = Pick<
+  InputHTMLAttributes<HTMLInputElement>,
+  "pattern" | "required" | "minLength" | "maxLength"
+>;
+
+/**
+ * Since the default validation messages can be verbose, this type is used to
+ * configure when/how to display the native browser messages when the validation
+ * state changes during the `change` event phase. The validation message will
+ * always be shown on blur.
+ *
+ * When this is:
+ *
+ * - `true` -&gt; always show the browser message when it exists
+ * - `false` -&gt; never show the browser message
+ * - `"recommended"` -&gt; only shows the browser message if it is not one of the
+ *   `RECOMMENDED_IGNORED_KEYS` validation errors
+ * - `keyof ValidityState` -&gt; only shows the browser message if it is not the
+ *   specific validation error
+ * - `(keyof ValidityState)[]` -&gt; only shows the browser message if it is not
+ *   the specific validation errors
+ *
+ * @remarks \@since 2.5.0
+ */
+export type ChangeValidationBehavior =
+  | boolean
+  | "recommended"
+  | "number-recommended"
+  | keyof ValidityState
+  | readonly (keyof ValidityState)[];
+
+/**
+ * @remarks \@since 2.5.0
+ */
+export interface ErrorMessageOptions extends TextConstraints {
+  /**
+   * The current input or textarea's validity state.
+   */
+  validity: ValidityState;
+
+  /**
+   * The browser defined validation message based on the validity state. This
+   * will be the empty string when there are no errors.
+   */
+  validationMessage: string;
+
+  /**
+   * The current `TextField` or `TextArea` value.
+   */
+  value: string;
+
+  /**
+   * Boolean if this is triggered from a blur event instead of a change event.
+   */
+  isBlurEvent: boolean;
+
+  /**
+   * The change event validation behavior that is specified in the hook.
+   */
+  validateOnChange: ChangeValidationBehavior;
+}
+
+/**
+ * A function to get a custom error message for specific errors. This is really
+ * useful when using the `pattern` attribute to give additional information or
+ * changing the native "language translated" error message.
+ *
+ * @param options - An object containing metadata that can be used to create an
+ * error message for your `TextField` or `TextArea`.
+ * @returns An error message to display or an empty string.
+ * @remarks \@since 2.5.0
+ */
+export type GetErrorMessage = (options: ErrorMessageOptions) => string;
+
+/**
+ * @internal
+ * @remarks \@since 2.5.0
+ */
+const VALIDITY_STATE_KEYS: readonly (keyof ValidityState)[] = [
+  "badInput",
+  "customError",
+  "patternMismatch",
+  "rangeOverflow",
+  "rangeUnderflow",
+  "stepMismatch",
+  "tooLong",
+  "tooShort",
+  "typeMismatch",
+  "valueMissing",
+];
+
+/**
+ * @internal
+ * @remarks \@since 2.5.0
+ */
+export const RECOMMENDED_STATE_KEYS: readonly (keyof ValidityState)[] = [
+  "badInput",
+  "tooLong",
+  "valueMissing",
+];
+
+/**
+ * @internal
+ * @remarks \@since 2.5.0
+ */
+export const RECOMMENDED_NUMBER_STATE_KEYS: readonly (keyof ValidityState)[] = [
+  ...RECOMMENDED_STATE_KEYS,
+  "rangeOverflow",
+  "rangeUnderflow",
+  "tooShort",
+  "typeMismatch",
+];
+
+/**
+ * The validation message is actually kind of weird since it's possible for a
+ * form element to have multiple errors at once. The validation message will be
+ * the first error that appears, so need to make sure that the first error is
+ * one of the recommended state keys so the message appears for only those types
+ * of errors.
+ *
+ * @internal
+ * @remarks \@since 2.5.0
+ */
+const isRecommended = (validity: ValidityState, isNumber: boolean): boolean => {
+  const errorable = isNumber
+    ? RECOMMENDED_NUMBER_STATE_KEYS
+    : RECOMMENDED_STATE_KEYS;
+
+  return VALIDITY_STATE_KEYS.every((key) => {
+    const errored = validity[key];
+    return !errored || errorable.includes(key);
+  });
+};
+
+/**
+ * The default implementation for getting an error message for the `TextField`
+ * or `TextArea` components that relies on the behavior of the
+ * {@link ChangeValidationBehavior}
+ *
+ * @remarks \@since 2.5.0
+ */
+export const defaultGetErrorMessage: GetErrorMessage = (options) => {
+  const { isBlurEvent, validity, validationMessage, validateOnChange } =
+    options;
+
+  if (isBlurEvent || !validationMessage || validateOnChange === true) {
+    return validationMessage;
+  }
+
+  if (!validateOnChange) {
+    return "";
+  }
+
+  if (
+    validateOnChange === "recommended" ||
+    validateOnChange === "number-recommended"
+  ) {
+    return isRecommended(validity, validateOnChange === "number-recommended")
+      ? validationMessage
+      : "";
+  }
+
+  const keys =
+    typeof validateOnChange === "string"
+      ? [validateOnChange]
+      : validateOnChange;
+
+  return keys.length &&
+    VALIDITY_STATE_KEYS.some((key) => validity[key] && keys.includes(key))
+    ? validationMessage
+    : "";
+};
+
+/**
+ * @remarks \@since 2.5.0
+ */
+export interface IsErroredOptions extends ErrorMessageOptions {
+  /**
+   * The current error message or an empty string.
+   */
+  errorMessage: string;
+}
+
+/**
+ * A function that is used to determine if a `TextField` or `TextArea` is in an
+ * errored state.
+ *
+ * @param options - All the current options that can be used to determine the
+ * error state.
+ * @returns True if the component is considered to be in an errored state.
+ * @remarks \@since 2.5.0
+ */
+export type IsErrored = (options: IsErroredOptions) => boolean;
+
+/**
+ * The default implementation for checking if a `TextField` or `TextArea` is
+ * errored by returning `true` if the `errorMessage` string is truthy or the
+ * value is not within the `minLength` and `maxLength` constraints when they
+ * exist.
+ *
+ * @remarks \@since 2.5.0
+ */
+export const defaultIsErrored: IsErrored = (options) => {
+  const { value, errorMessage, minLength, maxLength, isBlurEvent } = options;
+
+  return (
+    !!errorMessage ||
+    (typeof maxLength === "number" && value.length > maxLength) ||
+    (isBlurEvent && typeof minLength === "number" && value.length < minLength)
+  );
+};
+
+/**
+ * @remarks \@since 6.0.0
+ */
+export interface GetErrorIconOptions {
+  /**
+   * This will be `true` if the `TextField` or `TextArea` is in an errorred state.
+   */
+  error: boolean;
+
+  /**
+   * The current error icon that was provided.
+   */
+  errorIcon: ReactNode;
+
+  /**
+   * The current error message or an empty string.
+   */
+  errorMessage: string;
+}
+
+/**
+ * A function that can be used to dynamically get an error icon based on the
+ * current visible error.
+ *
+ * @param options - The {@link GetErrorIconOptions}
+ * @returns An icon to render or falsey to render nothing.
+ * @remarks \@since 2.5.0
+ * @remarks \@since 6.0.0 Updated to accept a single object argument
+ */
+export type GetErrorIcon = (options: GetErrorIconOptions) => ReactNode;
+
+/**
+ * The default implementation for showing an error icon in `TextField` and
+ * `TextArea` components that will only display when the error flag is enabled.
+ *
+ * @remarks \@since 2.5.0
+ */
+export const defaultGetErrorIcon: GetErrorIcon = (options) => {
+  const { error, errorIcon } = options;
+
+  return error && errorIcon;
+};
diff --git a/src/components/TextFieldValidation.tsx b/src/components/TextFieldValidation.tsx
new file mode 100644
index 000000000..e82d9a9c1
--- /dev/null
+++ b/src/components/TextFieldValidation.tsx
@@ -0,0 +1,124 @@
+import { Button } from "@react-md/button";
+import {
+  defaultGetErrorMessage,
+  ErrorChangeHandler,
+  Form,
+  GetErrorMessage,
+  TextField,
+  useTextField,
+} from "@react-md/form";
+import { ReactElement, useCallback, useMemo, useState } from "react";
+
+type FieldId = string;
+type ErrorRecord = Record<FieldId, boolean | undefined>;
+
+/**
+ * You can define a custom function to get the error message as well.
+ * @see defaultGetErrorMessage for more details
+ */
+const getErrorMessage: GetErrorMessage = (options) => {
+  const { validity, minLength, maxLength } = options;
+  if (validity.tooLong) {
+    return `No more than ${maxLength} characters.`;
+  }
+
+  if (validity.tooShort) {
+    return `Must be at least ${minLength} characters.`;
+  }
+
+  if (validity.valueMissing) {
+    return "This value is required!";
+  }
+
+  return defaultGetErrorMessage(options);
+};
+
+export default function TextFieldValidation(): ReactElement {
+  const [errors, setErrors] = useState<ErrorRecord>({});
+  const errored = useMemo(() => Object.values(errors).some(Boolean), [errors]);
+  const onErrorChange = useCallback<ErrorChangeHandler>(
+    (id, error) => setErrors((prevErrors) => ({ ...prevErrors, [id]: error })),
+    []
+  );
+
+  const { fieldProps: nameFieldProps } = useTextField({
+    required: true,
+    onErrorChange,
+  });
+  const { fieldProps: dateFieldProps } = useTextField({
+    required: true,
+    helpText: "mm/dd/yyyy",
+    pattern: "^\\d{2}/\\d{2}/\\d{4}$",
+    minLength: 10,
+    maxLength: 10,
+    onErrorChange,
+  });
+  const { fieldProps: descriptionFieldProps } = useTextField({
+    counter: true,
+    maxLength: 200,
+    // just so you can see the error state since browser prevent text from being
+    // entered when the maxLength attribute is provided
+    disableMaxLength: true,
+    onErrorChange,
+  });
+  const { fieldProps: description2FieldProps } = useTextField({
+    required: true,
+    counter: true,
+    minLength: 10,
+    maxLength: 50,
+    onErrorChange,
+    // disable the error icon, this is also pulled automatically from the
+    // `IconProvider` as the `error` icon when undefined
+    errorIcon: null,
+    // custom error messages for this one
+    getErrorMessage,
+  });
+  const { fieldProps: passwordProps } = useTextField({
+    required: true,
+    minLength: 10,
+    helpText: "Create a password with at least 10 characters.",
+    onErrorChange,
+  });
+
+  return (
+    <Form>
+      <TextField
+        {...nameFieldProps}
+        label="Name"
+        placeholder="John Doe"
+        name="name"
+      />
+      <TextField
+        {...dateFieldProps}
+        label="Date *"
+        placeholder="01/01/2020"
+        name="date"
+      />
+      {/* <TextArea */}
+      {/*   {...descriptionFieldProps} */}
+      {/*   label="Description" */}
+      {/*   placeholder="Something amazing." */}
+      {/*   name="description" */}
+      {/* /> */}
+      {/* <TextArea */}
+      {/*   {...description2FieldProps} */}
+      {/*   label="Description 2 *" */}
+      {/*   placeholder="Something amazing." */}
+      {/*   name="description2" */}
+      {/* /> */}
+      {/* <Password */}
+      {/*   {...passwordProps} */}
+      {/*   label="Password *" */}
+      {/*   name="password" */}
+      {/* /> */}
+      <Button
+        type="submit"
+        disabled={errored}
+        theme="primary"
+        themeType="outline"
+      >
+        Submit
+      </Button>
+    </Form>
+  );
+}
